### PR TITLE
Copter: fixed parachute checks for sink rate

### PR DESCRIPTION
**Pull request by tridge**
ensure is_flying is set, setup sink rate at the right position, force
disarm if chute releases

### CODE REVIEW

File: ArduCopter/crash_check.cpp [hunk 1]
```diff
@@ -209,6 +214,9 @@ void Copter::parachute_check()
         return;
     }
 
+    // pass sink rate to parachute library
```
Comment (tridge) [inline ArduCopter/crash_check.cpp:217]: maybe this should go above update() ?

**Comment by brentmit2010** [comment]
Just chiming in as this PR is due to the conversation we had earlier on the Partner call. We ran these (nearly identical) changes through 50+ parachute deployments with great success.

Don't hesitate to reach out if I can help in any way!

File: ArduCopter/crash_check.cpp [hunk 2]
```diff
@@ -192,6 +192,11 @@ void Copter::parachute_check()
         return;
     }
 
+    if (parachute.release_initiated()) {
```
Comment (khancyr) [inline ArduCopter/crash_check.cpp:195]: why do we need that ? 
the only way to have _release_initiated = true; is to call release() that pass on parachute_release() that issue the disarm already. 
So I don't understant why we need to spam disarm command ?

File: ArduCopter/crash_check.cpp (hunk 1 above)
Comment (khancyr) [reply ArduCopter/crash_check.cpp:217]: yep, looks better before the update()

File: ArduCopter/Copter.cpp [hunk 3]
```diff
@@ -512,6 +512,10 @@ void Copter::one_hz_loop()
     adsb.set_is_flying(!ap.land_complete);
 #endif
 
+#if PARACHUTE == ENABLED
+    parachute.set_is_flying(!ap.land_complete);
```
Comment (khancyr) [inline ArduCopter/Copter.cpp:516]: totally missed from https://github.com/ArduPilot/ardupilot/pull/9506/files so nice to have it !

File: ArduCopter/crash_check.cpp (hunk 2 above)
Comment (brentmit2010) [reply ArduCopter/crash_check.cpp:195]: The issue we ran into was that the sink rate check is done from the parachute library itself ( https://github.com/tridge/ardupilot/blob/pr-copter-chute-flying/libraries/AP_Parachute/AP_Parachute.cpp#L128 ) and not here in `crash_check` so disarm was never commanded.

Looking again at `parachute_release()` we probably want to include the landing gear deployment here, too?

```
    // deploy landing gear
    landinggear.set_position(AP_LandingGear::LandingGear_Deploy);
```

Comment (khancyr) [reply ArduCopter/crash_check.cpp:195]: Ok I have look back a little more that won't be and issue to use this. It won't spam.

About the landinggear, that can be a touchy change as some will prefer to not try to deploy back the landing gear (possible interference with the parachute cable, you may touch the ground before deployement, etc etc.). Better make another PR to put the landing gear change. Doing Mount retract could be a good things on other matter !

**Comment by rmackay9** [comment]
There's an edge case that can lead to a false positive release because the parachute.set_sink_rate() is only called about half way down in the parachute_check() function after seven potential early returns.  Because AP_Parachute does not check how old the sink rate is a single large sink rate could trigger a parachute release if it happened a moment before one of the early returns.  So for example, if the vehicle fell rapidly and then the user switched the vehicle to Acro mode or if a hard bang on the ground caused a large sink rate and then the vehicle was soon after considered landed or disarmed it could trigger... it is unlikely however.

What makes this messy is the triggering logic is split across AP_Parachute's update function and the vehicle code.  The easiest way to resolve this edge case is to move the critical sink rate check to a separate function and then:
 - Plane can call the new function right before parachute.update()
 - Copter can call the function after all the early returns (or it could simply ignore this function and instead integrate the equivalent into it's existing release logic)

I could try and fix this is people would like..

Two minor issues found with AP_Parachute:
 - I think the critical sink logic doesn't obey the alt-min in Plane.
 - critical_sink() is never used

**Comment by brentmit2010** [comment]
Ah, good catch on the edge case! For our particular use case, we don't use Acro and we have a `min_alt` set for the parachute, so I'm not _too_ concerned about it. But I can definitely see the hesitation on the current implementation being used more broadly.

I had similar headaches when trying to update the logic on our side with it being split between vehicle and parachute code. I think there is value in fixing that if you're up for it. :)

**Comment by rmackay9** [comment]
I've made my changes on top of this PR in this new branch https://github.com/rmackay9/rmackay9-ardupilot/commits/copter-parachute-fix.  I haven't tested it yet but I suspect it will work.  Any feedback is greatly appreciated.

**Comment by rmackay9** [comment]
Closing this in favour of a new PR https://github.com/ArduPilot/ardupilot/pull/15473

**Pull request closed without merging**